### PR TITLE
[FLINK-24394][test] Refactor BuiltInFunctions IT Tests

### PR DESCRIPTION
**Pull request by matriv**
Add the possibility to tests multiple TableApi expressions
and/or SQL expressions as multiple table columns using just one
execution of the pipeline to speed up the total test execution time.

Refactor existing tests of BuiltInFunctions to use the new approach
wherever possible.

## What is the purpose of the change

Refactor `BuiltInFunctionTestBase` to allow testing of multiple expressions at once.

## Brief change log

  - Add the possibility to test multiple TableApi and SQL expressions within on execution, by converting them into multiple table columns. This allows for faster overall test execution.
  - Refactor existing builtin function tests to use this new approach
  - Change some of the existing tests so that both Table API and SQL paths are verified.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by matriv** [comment]
@flinkbot run azure

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:258]: nit: could we use `List` instead of arrays? I find `Arrays.asList` and soon `List.of` nicer than the verbose array syntax.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:302]: nit: we don't enforce using final in Flink but if you check the core, you will see that most committers use final extensively to indicate immutable/mutability in code. I would vote for having a conistent coding style within our team. At least we should adapt to the coding style of the class. When I read this and the following lines, it looks to me as if we would modify the reference in the for loop again. Because all variables above are marked `final`.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:343]: nit: use the Java streams API or `String.join` once we have lists

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:489]: if Table API and SQL API share the same result it should in most cases also have the same data type?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:468]: `Column` sounds very internal to this class. Tests don't know columns but only test specs. How about `ResultSpec` to keep it short?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java:44]: instead of using a static import in every test, we could simply provide static methods in the upper class, called `resultItem(...)`

**Review by twalthr — COMMENTED** [review]
Thanks for the PR @matriv. This is quite a big change and should have a corresponding JIRA issue maybe as a subtask of your current issue. I added some feedback.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:258]: Don't have any objection, just chose arrays here to avoid the overhead of List objects especially for the common cases of only one element, but considering it's test, Lists are nicer indeed.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:489]: I kept this variant exposed as there are a couple of tests in `JsonFunctionsITCase` that make use of different return data types: https://github.com/apache/flink/pull/17341/files#diff-14f6e87fed00e28afef4a5e995126c430b4065661e3dc2f66d14f3741378f874R180

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:468]: Makes sense, thx, couldn't come up with a nice name.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CoalesceFunctionITCase.java:44]: Sure, I'd prefer though the name `resultSpec()` since `resultItem` is  another more lower level pojo within the class and can cause confusion.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:155]: `of spec` btw I think it would be nice to also print the test item summary to quickly know which one failed if there are many.

**Review by twalthr — APPROVED** [review]
Thanks for the update. I added one last comment and will merge this once the build is green then.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (twalthr) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java]: TBH this is a very special case and hopefully does not happen elsewhere. we should not start having more of these special cases.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java
Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java:155]: Something like:
```
"Logical type for spec [" + i + "] of test [" + testItem + "] doesn't match"
&
"Result for spec [" + i + "] of test [" + testItem + "] doesn't match"
```

Comment (matriv) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/BuiltInFunctionTestBase.java]: Pushed this change, along with a fix regarding the relevant `toString()` implementation for TableAPI cases.

**Comment by twalthr** [comment]
@flinkbot run azure

**Comment by matriv** [comment]
@twalthr I guess this should go to 1.14 as well and possibly 1.13 to ease future backporting of relevant fixes?

**Comment by twalthr** [comment]
@matriv I haven't merged this to 1.14 yet. We can do this on demand.

**Pull request closed without merging**
